### ISSUE TITLE
Group by arbitrary prop (including ref), return top k per group

### ISSUE THREAD

**Issue by etiennedi** [weaviate/weaviate#2602]
## Motivation

Assume you have a `Document<>Passage` schema where a document typically has many (e.g., 1000s of) Passages. Vector pooling methods such as `ref2vec` are problematic in this case because it's difficult to represent 1000s of diverse passages in a single vector. Some meaning is bound to get lost. So ideally the actual search should happen on individual passages, while still being able to make statements about `Documents`. For example "Considering the top 10k passages, what are the 10 best documents and for each document, what are the 5 passages closest to my query?". 

### Tech terms
Essentially we want to do a vector search first to retrieve `limit` candidates. This could be any search, but sticking with vector search for simplicity. Then aggregate the candidates by a grouping prop, while keeping at most `limit_per_group` elements per group and keeping at most `groups_limit` groups. 

### Specify how to group

```graphql
{
  Get{
    Passage(
      limit: 1000
      groupBy:{
        path: ["ofDocument"]
        groups: 10,
        objectPerGroup: 10
      }
    ) {
      content
      ofDocument{
        ... on Document {
          title
        }
      }
    }
  }
}
```

#### Notes

* Note that the value for `path` is a ref Prop, however, the path has `length=1`. This is because performance is a big concern for this proposal. Grouping by a ref's ID does not require resolving the ref. However, if the path was `["ofDocument", "Document", "title"]` this would require resolving all documents and would come with a considerably higher price tag. 
* This section only specifies how to control the grouping, but not how to view the groups. Since I have two alternatives, I have split this into a separate section below.

### Returning Results

The following are two alternatives to show how results belong to a group. One makes use of a nested structure, the other keeps a flat list.

#### Alternative 1 - "Flat"

**IMPORTANT: This proposal was rejected in favor of the other one**

The idea behind Alternative 1 is that the return structure stays flat and group association is done purely through an additional prop that indicates which group a prop belongs to.

Since this is a flat list, we would have all elements in the list. So with `groups=10` and `objectsPerGroup=10` we would have up to 100 results. (Fewer results are possible if not enough candidates were returned to fill all 10 groups, in this case the user needs to increase `limit`).

##### Request
```graphql
{
  Get{
    Passage(
      limit: 1000
      groupBy:{
        path: ["ofDocument"]
        groups: 3,
        objectPerGroup: 3
      }
    ) {
      content
      ofDocument{
        ... on Document {
          title
        }
      }
      _additional {
        distance
        groupBy {
          groupId     # these are just examples for group metadata
          minDistance # these are just examples for group metadata
          maxDistance # these are just examples for group metadata
          count       # these are just examples for group metadata
        }
      }
    }
  }
}
```

##### Response

```yaml
"Passage": [
  {
    "content": "Content of passage 1",
    "ofDocument": {
      "title": "Title of Document A",
    }
    "_additional": {
       "distance": 0.1
       "groupBy": {
          "groupId": 0
          "minDistance": 0.1,
          "maxDistance": 0.3,
          "count": 3,
       }
    }
  },
  {
    "content": "Content of passage 2",
    "ofDocument": {
      "title": "Title of Document A",
    }
    "_additional": {
       "distance": 0.2
       "groupBy": {
          "groupId": 0
          "minDistance": 0.1,
          "maxDistance": 0.3,
          "count": 3,
       }
    }
  },
 {
    "content": "Content of passage 3",
    "ofDocument": {
      "title": "Title of Document A",
    }
    "_additional": {
       "distance": 0.3
       "groupBy": {
          "groupId": 0
          "minDistance": 0.1,
          "maxDistance": 0.3,
          "count": 3,
       }
    }
  },
 {
    "content": "Content of passage 4",
    "ofDocument": {
      "title": "Title of Document B," # <--- NOTE THAT THIS IS A NEW GROUP
    }
    "_additional": {
       "distance": 0.21 # <--- lowest distance of group two, but not lowest distance overall!
       "groupBy": {
          "groupId": 1  # <--- NOTE THAT THIS IS A NEW GROUP
          "minDistance": 0.21,
          "maxDistance": 0.7,
          "count": 3,
       }
    }
  },
  # etc...
]
```

#### Alternative 2 - "Nested"

**IMPORTANT: This proposal was accepted, not the other one**

The idea here is that what's returned is not all 100 objects, but rather just the 10 groups. Then, using an `_additional` prop one can access the individual _inner_ objects:

##### Request
```graphql
{
  Get{
    Passage(
      limit: 1000
      groupBy:{
        path: ["ofDocument"]
        groups: 3,
        objectPerGroup: 3
      }
    ) {
      ofDocument{ # <-- at the root level we can only specify prop that's used to group, as the root level is no longer a real object, it's just a wrapper to indicate the right group
        ... on Document {
          title
        }
      }
      _additional {
        distance
        groupBy {
          groupId     # these are just examples for group metadata
          minDistance # these are just examples for group metadata
          maxDistance # these are just examples for group metadata
          count       # these are just examples for group metadata
        }
        groupedHits {  # <-- this is where the actual passages for each document will be
          content
          _additional { _distance }
        }
      }
    }
  }
}
```

##### Response

```yaml
"Passage": [
  {
    "ofDocument": { # the outer objects can only contain grouped props, all other props would be `null` as the other object does not exist, it's just a fictional wrapper to group
      "title": "Title of Document A",
    }
    "_additional": {
       "groupBy": {
          "groupId": 0
          "minDistance": 0.1,
          "maxDistance": 0.3,
          "count": 3,
       }
      "groupedHits": [{
          "content": "Content of "passage 1",
          "_additional": {
            "distance": 0.1
          } 
       },{
          "content": "Content of "passage 2",
          "_additional": {
            "distance": 0.2
          } 
       },{
          "content": "Content of "passage 3",
          "_additional": {
            "distance": 0.3
          } 
       }]
    }
  },
 {
    "ofDocument": { # the outer objects can only contain grouped props, all other props would be `null` as the other object does not exist, it's just a fictional wrapper to group
      "title": "Title of Document B",
    }
    "_additional": {
       "groupBy": {
          "groupId": 0
          "minDistance": 0.21,
          "maxDistance": 0.7,
          "count": 3,
       }
      "groupedHits": [{
          "content": "Content of "passage 4",
          "_additional": {
            "distance": 0.21
          } 
       },
        # etc
      ]
    }
  },
```

## Request for feedback

Before diving deeper, I'm curious about:

* What do you think about this in general?
* Do you have a preference for the flat or nested API?

**Comment by kcm** [comment]
Approach 2 looks more straightforward to develop around. While this would be abstracted out somehow with a language client, it would be a bit more performant to do client-side calculations with the grouping arrangement there.

**Comment by etiennedi** [comment]
After getting some internal and external feedback it indeed looks like Approach 2 (nested) is the way to go.

**Comment by antas-marcin** [comment]
After further discussions we agreed to modify the structure of the `_addtional.group` parameter:

```graphQL
{
  Get{
    Passage(
      limit: 1000
      groupBy:{
        path: ["ofDocument"]
        groups: 3,
        objectPerGroup: 3
      }
    ) {
      ofDocument{ # <-- at the root level we can only specify prop that's used to group, as the root level is no longer a real object, it's just a wrapper to indicate the right group
        ... on Document {
          title
        }
      }
      _additional {
        distance
        group {
          id          # these are just examples for group metadata
          minDistance # these are just examples for group metadata
          maxDistance # these are just examples for group metadata
          count       # these are just examples for group metadata

          hits {  # <-- this is where the actual passages for each document will be
            content
            _additional { _distance }
          }
        }
      }
    }
  }
}
```

Response:

```
"Passage": [
  {
    "ofDocument": { # the outer objects can only contain grouped props, all other props would be `null` as the other object does not exist, it's just a fictional wrapper to group
      "title": "Title of Document A",
    }
    "_additional": {
       "group": {
          "id": 0
          "minDistance": 0.1,
          "maxDistance": 0.3,
          "count": 3,
          "hits": [{
              "content": "Content of "passage 1",
              "_additional": {
                "distance": 0.1
              } 
           },{
              "content": "Content of "passage 2",
              "_additional": {
                "distance": 0.2
              } 
           },{
              "content": "Content of "passage 3",
              "_additional": {
                "distance": 0.3
              } 
           }]
       }
    }
  },
 {
    "ofDocument": { # the outer objects can only contain grouped props, all other props would be `null` as the other object does not exist, it's just a fictional wrapper to group
      "title": "Title of Document B",
    }
    "_additional": {
       "group": {
          "id": 0
          "minDistance": 0.21,
          "maxDistance": 0.7,
          "count": 3,
          "hits": [{
              "content": "Content of "passage 4",
              "_additional": {
                "distance": 0.21
              } 
           },
            # etc
          ]
       }
    }
  },
```

**Comment by nairajay2k** [comment]
I want a document list (paginated preferably) ordered by a custom score which is the weighted average of top 3 paragraph scores in the document. Its not totally clear to me if the feature on this page can solve that. If yes , please do let know by when this could be available.

Thanks

**Comment by etiennedi** [comment]
Hi @nairajay2k and thank you for your interest in this feature. In the current proposal and first iteration a groups score (and therefore position) is determined by the first (closest) element in the group. 

But I'm sure this could be extended to use other weighting mechanisms, such as the mean score of the group or the top n elements of a group. Please feel free to open a dedicated issue for this as we will close this issue once the first iteration is ready and released.

**Comment by nairajay2k** [comment]
Hi @etiennedi , opened a new issue .

**Comment by etiennedi** [comment]
Released in v1.19.
